File: ps/leetcode_cpp/MinStack.cpp
#include <iostream>
#include <stack>
#include <string>
#include <map>
using namespace std;


class MinStack {
private:
    stack<int> stk;
    map<int, int> mp;
public:
    MinStack() {
        
    }
    
    void push(int x) {
        stk.push(x);
        ++mp[x];
    }
    
    void pop() {
        int x = stk.top();
        stk.pop();
        --mp[x];
        if(mp[x] == 0)
            mp.erase(x);
    }
    
    int top() {
        return stk.top();
    }
    
    int getMin() {
        return mp.begin()->first;
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(x);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */

File: ps/leetcode_java/src/MinStack/MinStack.java
package MinStack;

import java.util.Stack;

public class MinStack {
    class Pair{
        int val, min;
        Pair(){}
        Pair(int val, int min) {
            this.val = val;
            this.min = min;
        }
    }

    Stack<Pair> stk;
    public MinStack() {
        stk = new Stack<>();
    }

    public void push(int x) {
        int min;
        if(stk.empty()) min = x;
        else{
            min = Math.min(x, stk.peek().min);
        }
        stk.push(new Pair(x, min));
    }

    public void pop() {
        stk.pop();
    }

    public int top() {
        return stk.peek().val;
    }

    public int getMin() {
        return stk.peek().min;
    }
}
